src/tree.rs

```diff
@@ -1423,6 +1423,88 @@ impl<T> Tree<T> {
     }
 }
 
+impl<T> PartialEq for Tree<T>
+where
+    T: PartialEq,
+{
+    fn eq(&self, other: &Tree<T>) -> bool {
+        if self.nodes.iter().filter(|x| x.is_some()).count()
+            != other.nodes.iter().filter(|x| x.is_some()).count()
+        {
+            return false;
+        }
+
+        for ((i, node1), (j, node2)) in self
+            .nodes
+            .iter()
+            .enumerate()
+            .filter_map(|(i, x)| (*x).as_ref().map(|x| (i, x)))
+            .zip(
+                other
+                    .nodes
+                    .iter()
+                    .enumerate()
+                    .filter_map(|(i, x)| (*x).as_ref().map(|x| (i, x))),
+            )
```

Comment:
Question: do you know why we have a vector of Option for nodes in Tree instead of just having a vector of Node? I'm questioning all the logic about the free_ids too here.

We won't need the two `.filter_map(|(i, x)| (*x).as_ref().map(|x| (i, x)))` if we had a Vec instead of Vec<Option>